### ISSUE TITLE
out_of_bounds_indexing false positive

### ISSUE THREAD

**Issue by leonardo-m** [Manishearth/rust-clippy#1668]
This comes from a code reduction, so it's reasonable in its original context:

```
#![allow(unused_variables)]
fn main() {
    const N: usize = 1_000;
    let data = [0, 0, 0, 0];
    if N <= 4 {
        let result = data[N - 1];
    }
}```


```
error: const index is out of bounds
 --> src\main.rs:6:22
  |
6 |         let result = data[N - 1];
  |                      ^^^^^^^^^^^
  |
  = note: #[deny(out_of_bounds_indexing)] on by default
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#out_of_bounds_indexing
```

**Comment by felix91gr** [comment]
This bug still stands: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=47c984fabc4858ae2b655e9ce7f5c4d0

@oli-obk this issue seems very closely related to const eval, which you're (last I checked) a champion of. Is there anything in Clippy or the compiler that could make this lint's analysis more refined?

Though I think counterexamples of this kind might still exist until we get something comparable to Value-Range Analysis.

---

Regardless, perhaps this issue could use more tags? I'd give it at least `T-MIR` and `I-False-Positive`. Maybe also `S-Blocked` if it turns out that we're out of luck until MIR's Dataflow Analysis becomes capable of reasoning about these kinds of contexts.

**Comment by Jarcho** [comment]
Doing value range and known bits analyses is a thing I would like to implement in clippy at some point, but it won't be anytime soon.

**Comment by felix91gr** [comment]
> Doing value range and known bits analyses is a thing I would like to implement in clippy at some point, but it won't be anytime soon.

Same. The underlying lattices required for VRA are gargantuan compared to the ones required for constprop. Which are themselves a lot harder to work with than the lattices required for bitfields. And currently the compiler can only do dataflow analysis on bitfields... so we're definitely a long ways away yet.
